### PR TITLE
Add gateway support for Starknet 0.13.0

### LINKED ISSUE #1483: Support V3 transactions in `starknet_gateway_types`
Update the gateway's transaction schema to include the new V3 transaction types. ~The schema is not yet available.~ Schema examples below.

### PR DESCRIPTION
**Pull request by kkovaacs**
This PR adds support for Starknet 0.13.0 changes:
- v3 transaction types
- new `eth_l1_gas_price` and `strk_l1_gas_price` fields in blocks
- some fixes to P2P protobuf definitions

A few `todo!()`s are still there because:
- no conversion to `BroadcastedTransaction` is possible yet (no v3 support there yet)
- conversion to `starknet_api` transaction types is not yet possible (depends on upgrading `starknet_api` and `blockifier`).

These could both be added to this PR but I think this is getting already too large...

Closes #1483

### CODE REVIEW

File: crates/gateway-types/src/transaction_hash.rs [hunk 1]
```diff
@@ -506,6 +701,7 @@ mod tests {
             case!(v0_9_0::transaction::INVOKE),
             case!(v0_11_0::transaction::invoke::v1::BLOCK_420K),
             case!(v0_11_0::transaction::invoke::v1::BLOCK_790K),
+            case!(v0_13_0::transaction::invoke::v3::BLOCK_319106),
```
Comment (Mirko-von-Leipzig) [inline crates/gateway-types/src/transaction_hash.rs:704]: Non-PR note - consider replacing custom macro with `rstest`

File: crates/p2p/src/client/types.rs [hunk 2]
```diff
@@ -254,7 +256,51 @@ impl TryFromDto<p2p_proto::transaction::Transaction> for TransactionVariant {
...
                 compiled_class_hash: CasmHash(x.compiled_class_hash),
             }),
-            DeclareV3(_) => unimplemented!(),
+            DeclareV3(x) => TransactionVariant::DeclareV3(DeclareTransactionV3 {
+                class_hash: ClassHash(x.class_hash.0),
+                nonce: TransactionNonce(x.nonce),
+                nonce_data_availability_mode:
+                    pathfinder_common::transaction::DataAvailabilityMode::L1,
+                fee_data_availability_mode:
+                    pathfinder_common::transaction::DataAvailabilityMode::L1,
```
Comment (Mirko-von-Leipzig) [inline crates/p2p/src/client/types.rs:265]: Do these fields just not exist yet?

File: crates/common/src/transaction.rs
Comment (Mirko-von-Leipzig) [inline crates/common/src/transaction.rs:1]: Some blank lines in the variant structs definitions that should be removed.

File: crates/common/src/transaction.rs [hunk 3]
```diff
@@ -123,3 +203,22 @@ pub enum EntryPointType {
...
+    pub max_amount: ResourceAmount,
+    pub max_price_per_unit: ResourcePricePerUnit,
+}
+
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+#[repr(u8)]
+pub enum DataAvailabilityMode {
+    #[default]
+    L1 = 0,
+}
```
Comment (Mirko-von-Leipzig) [inline crates/common/src/transaction.rs:224]: Does this need `repr(u8)`?

File: crates/storage/src/schema/revision_0045.rs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+use anyhow::Context;
+
+pub(crate) fn migrate(tx: &rusqlite::Transaction<'_>) -> anyhow::Result<()> {
+    tx.execute_batch(
+        r"
+ALTER TABLE block_headers RENAME COLUMN gas_price TO eth_l1_gas_price;
+ALTER TABLE block_headers ADD COLUMN strk_l1_gas_price BLOB DEFAULT NULL;
```
Comment (Mirko-von-Leipzig) [inline crates/storage/src/schema/revision_0045.rs:7]: Are both L1 gas prices?

File: crates/common/src/transaction.rs (hunk 3 above)
Comment (kkovaacs) [reply crates/common/src/transaction.rs:224]: Probably not -- I'll remove that.

File: crates/gateway-types/src/reply.rs [hunk 5]
```diff
@@ -389,6 +391,65 @@ pub mod transaction {
         }
     }
 
+    #[derive(
+        Clone,
+        Default,
+        Debug,
+        serde_repr::Deserialize_repr,
+        serde_repr::Serialize_repr,
```
[the remark this answers is not in the document]
Comment (kkovaacs) [reply crates/gateway-types/src/reply.rs:408]: I can remove that crate and implement Serialize and Deserialize manually.

File: crates/p2p/src/client/types.rs (hunk 2 above)
Comment (kkovaacs) [reply crates/p2p/src/client/types.rs:265]: They do -- this was just a shortcut. I've added proper TryFromDto implementation for DataAvailabilityMode and use that here.

File: crates/storage/src/schema/revision_0045.rs (hunk 4 above)
Comment (kkovaacs) [reply crates/storage/src/schema/revision_0045.rs:7]: I guess yes. And the name of the filed is `strk_l1_gas_price` in the JSON from the feeder gateway.

File: crates/common/src/transaction.rs [hunk 6]
```diff
@@ -123,3 +188,21 @@ pub enum EntryPointType {
...
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+pub struct ResourceBound {
+    pub max_amount: ResourceAmount,
+    pub max_price_per_unit: ResourcePricePerUnit,
+}
+
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+pub enum DataAvailabilityMode {
+    #[default]
+    L1 = 0,
```
Comment (Mirko-von-Leipzig) [inline crates/common/src/transaction.rs:207]:
```suggestion
    L1,
```

File: crates/gateway-types/src/reply.rs [hunk 7]
```diff
@@ -389,6 +399,103 @@ pub mod transaction {
...
+    }
+
+    impl Serialize for DataAvailabilityMode {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: serde::Serializer,
+        {
+            serializer.serialize_u8(*self as u8)
+        }
+    }
```
Comment (Mirko-von-Leipzig) [inline crates/gateway-types/src/reply.rs:416]:
```suggestion
    #[derive(Copy, Clone, Default, Debug, PartialEq, Eq, Dummy)]
    pub enum DataAvailabilityMode {
        #[default]
        L1,
    }

    impl Serialize for DataAvailabilityMode {
        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
        where
            S: serde::Serializer,
        {
            match self {
                Self::L1 => serializer.serialize_u8(0),
            }
        }
    }
```

Comment (Mirko-von-Leipzig) [reply crates/gateway-types/src/reply.rs:416]: Why are they using integers here in any case 🤔 weird decision

Comment (kkovaacs) [reply crates/gateway-types/src/reply.rs:416]: It is -- but on the other hand they're using the integer representation of these values in the transaction hash computation so there has to be a canonical representation for them.

File: crates/common/src/transaction.rs (hunk 6 above)
Comment (kkovaacs) [reply crates/common/src/transaction.rs]: Done.

File: crates/gateway-types/src/reply.rs [hunk 8]
```diff
@@ -17,8 +17,11 @@ pub struct Block {
     pub block_number: BlockNumber,
     /// Excluded in blocks prior to Starknet 0.9
     #[serde_as(as = "Option<GasPriceAsHexStr>")]
-    #[serde(default)]
-    pub gas_price: Option<GasPrice>,
+    #[serde(default, alias = "gas_price")]
+    pub eth_l1_gas_price: Option<GasPrice>,
+    #[serde_as(as = "Option<GasPriceAsHexStr>")]
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub strk_l1_gas_price: Option<GasPrice>,
```
Comment (Mirko-von-Leipzig) [inline crates/gateway-types/src/reply.rs:24]: I assume they've kept the original field names in historic blocks?

**Review by Mirko-von-Leipzig — APPROVED** [review]

File: crates/gateway-types/src/reply.rs (hunk 8 above)
Comment (Mirko-von-Leipzig) [reply crates/gateway-types/src/reply.rs:24]: Maybe add a note to remove the options once v0.13 is deployed everywhere.

Comment (kkovaacs) [reply crates/gateway-types/src/reply.rs:25]: I've added a TODO.

**Pull request merged**